64jam.p8

```diff
@@ -126,82 +131,81 @@ function update_projectiles()
 	end
 end
 
-function fire_rocket()
-	if (player.rockets > 0) then
-		local x = 0
-		local y = 0
+-- converts a direction and speed to a dx, dy
+function dir_to_deltas(dir, speed)
+	local dx = 0
+	local dy = 0
+
+	if dir == 12 or dir == 10 or dir == 1 then
+		dy = -speed
+	end
+	
+	if dir == 5 or dir == 6 or dir == 7 then
+		dy = speed
+	end
+	 
+	if dir >= 1 and dir < 6 then
+		dx = speed
+	end
+	if dir > 6 and dir <= 10 then
+		dx = -speed
+	end 
+	return dx, dy
+end
+
+function create_projectile(speed)
+		local dx
+		local dy
 		local speed = 6
-		sfx(2)
 		
-		if player.dir == 12 or player.dir == 10 or player.dir == 1 then
-			y = -speed
-		end
-		
-		if player.dir == 5 or player.dir == 6 or player.dir == 7 then
-		y = speed
-		end
-		 
-		if player.dir >= 1 and player.dir < 6 then
-			x = speed
-		end
-		if player.dir > 6 and player.dir <= 10 then
-			x = -speed
-		end 
-		 
-		local r={
-			sp = 3,
+		dx, dy = dir_to_deltas(player.dir, speed)
+
+		local r= {
 			x = player.x,
 			y = player.y,
-			dx = x,
-			dy = y
+			dir = player.dir,
+			dx = dx,
+			dy = dy
 		}
+		return r
+end
+
+function fire_rocket()
+	if (player.rockets > 0) then
+		sfx(2)
+
+		local speed = 6
+		local r = create_projectile(speed)
+
 		add(rockets, r)
 		player.rockets -=1
 	end
 end
 
 function fire_bullet()
 	if (player.bullets > 0) then
-		local x = 0
-		local y = 0
-		local speed = 4
 		sfx (0)
-		if player.dir == 12 or player.dir == 10 or player.dir == 1 then
-			y = -speed
-		end		
-		if player.dir == 5 or player.dir == 6 or player.dir == 7 then
-		y = speed
-		end
-		if player.dir >= 1 and player.dir < 6 then
-			x = speed
-		end
-		if player.dir > 6 and player.dir <= 10 then
-			x = -speed
-		end
-		 
-		local b={
-			sp = 3,
-			x = player.x,
-			y = player.y,
-			dx = x,
-			dy = y
-		}
+
+		local speed = 4
+		local b = create_projectile(speed)
+
 		add(bullets, b)
 		player.bullets -=1
 	end
 end
 
 function draw_projectiles()
 	for r in all(rockets) do
-	 local s = 112
-	 if (player.dir == 12) then spr(115,r.x,r.y,1,1,false,true) end
-	 if (player.dir == 1) then spr(114,r.x,r.y,1,1,false,true)  end
-	 if (player.dir == 3) then spr(113,r.x, r.y) end
-	 if (player.dir == 5) then spr(114,r.x, r.y) end 
-	 if (player.dir == 6) then spr(115,r.x, r.y) end
-	 if (player.dir == 7) then spr(114,r.x,r.y,1,1,true,false) end
-	 if (player.dir == 9) then spr(113,r.x,r.y,1,1,true,true) end
-	 if (player.dir == 10) then spr(114,r.x,r.y,1,1,true,true) end
+		local s = 112
+		-- todo - standardise sprite layout, write "draw_sprite(ti, x, y, dir)"
```

Comment:
Yep, this function feels like it could be a one liner.  I need to sort out the sprite sheets soon (or our approach) as we're fast running out of room.